src/decoder/builder.rs

```diff
@@ -0,0 +1,307 @@
+//! let len = file.metadata().unwrap().len();
+//!
+//! let decoder = Decoder::builder()
+//!     .with_data(file)
+//!     .with_byte_len(len)      // Enable seeking and duration calculation
+//!     .with_hint("mp3")        // Optional format hint
+//!     .with_gapless(true)      // Enable gapless playback
+//!     .build()
+//!     .unwrap();
+//! ```
+//!
+//! # Settings
+//!
+//! The following settings can be configured:
+//!
+//! - `byte_len` - Total length of the input data in bytes
+//! - `hint` - Format hint like "mp3", "wav", etc
+//! - `mime_type` - MIME type hint for container formats
+//! - `seekable` - Whether seeking operations are enabled
+//! - `gapless` - Enable gapless playback
+//! - `coarse_seek` - Use faster but less precise seeking
+
+use std::io::{Read, Seek};
+
+#[cfg(feature = "symphonia")]
+use self::read_seek_source::ReadSeekSource;
+#[cfg(feature = "symphonia")]
+use ::symphonia::core::io::{MediaSource, MediaSourceStream};
+
+use super::*;
+
+/// Audio decoder configuration settings.
+/// Support for these settings depends on the underlying decoder implementation.
+/// Currently, settings are only used by the Symphonia decoder.
+#[derive(Clone, Debug)]
+pub struct Settings {
+    /// The length of the stream in bytes.
+    /// This is required for:
+    /// - Reliable seeking operations
+    /// - Duration calculations in formats that lack timing information (e.g. MP3, Vorbis)
+    ///
+    /// Can be obtained from file metadata or by seeking to the end of the stream.
+    pub(crate) byte_len: Option<u64>,
+
+    /// Whether to use coarse seeking.
+    /// Coarse seeking is faster but less accurate: it may seek to a position slightly before or
+    /// after the requested one, especially when the bitrate is variable.
+    pub(crate) coarse_seek: bool,
+
+    /// Whether to trim frames for gapless playback.
+    /// Note: Disabling this may affect duration calculations for some formats
+    /// as padding frames will be included.
+    pub(crate) gapless: bool,
+
+    /// An extension hint for the decoder about the format of the stream.
+    /// When known, this can help the decoder to select the correct codec.
+    pub(crate) hint: Option<String>,
+
+    /// An MIME type hint for the decoder about the format of the stream.
+    /// When known, this can help the decoder to select the correct demuxer.
+    pub(crate) mime_type: Option<String>,
+
+    /// Whether the decoder should report as seekable.
+    pub(crate) is_seekable: bool,
+}
+
+impl Default for Settings {
+    fn default() -> Self {
+        Self {
+            byte_len: None,
+            coarse_seek: false,
+            gapless: true,
+            hint: None,
+            mime_type: None,
+            is_seekable: false,
+        }
+    }
+}
+
+/// Builder for configuring and creating a decoder.
+///
+/// This provides a flexible way to configure decoder settings before creating
+/// the actual decoder instance.
+///
+/// # Examples
+///
+/// ```no_run
```

Comment:
rust documentation guidelines recommend to not use unwrap in example (users tend to copy them verbatim and keep them in). We copied that into our own [guidlines](https://github.com/RustAudio/rodio/blob/master/CONTRIBUTING.md#documentation)

In practice that means replace the `.unwrap` with `?` and start the example with:
`# fn main() -> Result<(), Box<dyn std::error::Error>>`
while ending with
`# Ok(())`